src/main/play-25/uk/gov/hmrc/play/it/MicroServiceEmbeddedServer.scala

```diff
@@ -0,0 +1,185 @@
+    val port = externalServicePorts.getOrElse(serviceName, throw new IllegalArgumentException(s"Unknown service '$serviceName'"))
+    s"http://localhost:$port/${-/(path)}"
+  }
+}
+
+trait EmbeddedServiceOrchestrator extends ResourceProvider with StartAndStopServer {
+
+  self: IntegrationTestConfiguration with ExternalServiceOrchestrator =>
+
+  import uk.gov.hmrc.play.it.SafelyStop._
+  import uk.gov.hmrc.play.it.UrlHelper._
+
+  protected val servicePort: Int = Port.randomAvailable
+
+  private lazy val server = {
+
+    val configMap = externalServicePorts.foldLeft(Map.empty[String, Any])(
+      (map, servicePort) => {
+        val serviceName = servicePort._1
+        val port = servicePort._2
+        Logger.debug(s"External service '$serviceName' is running on port: $port")
+
+        val updatedMap = map +
+          (s"$applicationMode.microservice.services.$serviceName.port" -> new Integer(port)) +
+          (s"$applicationMode.microservice.services.$serviceName.host" -> "localhost")
+
+        updatedMap
+      }) ++ additionalConfig
+
+    val configMapUpdated = onConfigUpdating(configMap).mapValues(_.asInstanceOf[AnyRef])
+    val config: Configuration = play.api.Configuration.from(configMapUpdated)
+
+
+    val environment: Environment = Environment.simple(mode = applicationMode)
+
+    val context = api.ApplicationLoader.createContext(environment, initialSettings = configMapUpdated)
+    val application = ApplicationLoader(context).load(context)
+
+    Play.start(application)
+
+    val serverConfig: ServerConfig = ServerConfig(rootDir = new File("."), port = Some(servicePort), address = "127.0.0.1")
+    val server = NettyServer.fromApplication(application, serverConfig)
+    addShutdownHook(server.stop)
+
+    server
+  }
+
+
+  // Give the tests a chance to access config before the application started
+  def onConfigUpdating(config: Map[String, Any]): Map[String, Any] = config
+
+  abstract override def start() {
+    super.start()
+    server
+  }
+
+  abstract override def stop() {
+    super.stop()
+    safelyStop("stopping Play server")(server.stop())
+  }
+
+  override def resource(url: String): String = s"http://localhost:$servicePort/${-/(url)}"
+}
+
+trait AdditionalConfigProvider {
+  protected def additionalConfig: Map[String, _ <: Any] = Map.empty
+}
+
+trait MongoTestConfiguration extends AdditionalConfigProvider {
+
+  self: IntegrationTestConfiguration =>
+
+  val dbName: String = testId.toString
+
+  abstract override protected def additionalConfig =
+    super.additionalConfig + (s"$applicationMode.microservice.mongodb.uri" -> s"mongodb://localhost:27017/$dbName")
+}
+
+object UrlHelper {
+  def -/(uri: String) = if (uri.startsWith("/")) uri.drop(1) else uri
+}
+
+private[it] object SafelyStop {
+
+  def safelyStop[T](activity: String)(action: => T) {
+    try {
+      action
+    } catch {
+      case t: Throwable => Logger.error(s"An exception occurred while $activity", t)
```

Comment:
Maybe we can catch non-fatal only? with `case NonFatal(t) => ...`

Reply:
That's probably a good idea, but I didn't want to make any functional changes to the library as part of this PR unless you're happy for me to do that.

Reply:
I don't think it would be harmful, but you're right, it is code that was already there :)